### PR TITLE
BUG: Fix bugs and performance issues in Segmentations

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
* Move and reformat "Binary labelmap layers" section of Segmentations module
* Only update layer section in Segmentations while it is open
* Add total segment count to layer section
* Show wait cursor while importing labelmap nodes to segmentation
* Improve performance when modifying segments through segment editor
* Fix bugs in Joint Smoothing and AutoComplete editor effects that allowed resulting segments to exist on multiple layers
* Improve 3D segment pipeline to improve speed when segment visibility is changed
* Preferred display representation is no longer set to be closed surface by default
* Fix bug that allowed segments to be added to the subject hierarchy twice

### CODE REVIEW

**Comment by Sunderlandkyl** [comment]
@lassoan This should fix most of the bugs mentioned here: https://github.com/Slicer/Slicer/pull/1221#issuecomment-539490163, except the test is still failing on MacOS.

I'll continue to look at that issue.

**Comment by Sunderlandkyl** [comment]
Performance wise, the only remaining factor seems to be copying the segments for undo/redo.

File: Modules/Loadable/Segmentations/EditorEffects/Python/AbstractScriptedSegmentEditorAutoCompleteEffect.py [hunk 1]
```diff
@@ -293,8 +293,14 @@ def onApply(self):
       segmentID = segmentIDs.GetValue(index)
       previewSegment = previewNode.GetSegmentation().GetSegment(segmentID)
       previewSegmentLabelmap = previewSegment.GetRepresentation(vtkSegmentationCore.vtkSegmentationConverter.GetSegmentationBinaryLabelmapRepresentationName())
-      self.scriptedEffect.modifySegmentByLabelmap(segmentationNode, segmentID,previewSegmentLabelmap,
+
+      oldOverwriteMode = self.scriptedEffect.parameterSetNode().GetOverwriteMode()
+      self.scriptedEffect.parameterSetNode().SetOverwriteMode(slicer.vtkMRMLSegmentEditorNode.OverwriteAllSegments)
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/AbstractScriptedSegmentEditorAutoCompleteEffect.py]: Why is this necessary?

Auto-complete effect ignores hidden segments, but now with this change, those segments would be modified, which is probably not what the user wants.

File: Modules/Loadable/Segmentations/EditorEffects/Python/AbstractScriptedSegmentEditorAutoCompleteEffect.py [hunk 2]
```diff
@@ -474,7 +480,12 @@ def preview(self):
         color = segmentationNode.GetSegmentation().GetSegment(segmentID).GetColor()
         newSegment.SetColor(color)
         previewNode.GetSegmentation().AddSegment(newSegment, segmentID)
+
+      oldOverwriteMode = self.scriptedEffect.parameterSetNode().GetOverwriteMode()
+      self.scriptedEffect.parameterSetNode().SetOverwriteMode(slicer.vtkMRMLSegmentEditorNode.OverwriteAllSegments)
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/AbstractScriptedSegmentEditorAutoCompleteEffect.py]: Why is this needed?

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorSmoothingEffect.py [hunk 3]
```diff
@@ -342,8 +342,12 @@ def smoothMultipleSegments(self):
       smoothedBinaryLabelMap.ShallowCopy(stencil.GetOutput())
       smoothedBinaryLabelMap.SetImageToWorldMatrix(imageToWorldMatrix)
       # Write results to segments directly, bypassing masking
-      slicer.vtkSlicerSegmentationsModuleLogic.SetBinaryLabelmapToSegment(smoothedBinaryLabelMap,
-        segmentationNode, segmentId, slicer.vtkSlicerSegmentationsModuleLogic.MODE_REPLACE, smoothedBinaryLabelMap.GetExtent())
+      oldOverwriteMode = self.scriptedEffect.parameterSetNode().GetOverwriteMode()
+      self.scriptedEffect.parameterSetNode().SetOverwriteMode(slicer.vtkMRMLSegmentEditorNode.OverwriteAllSegments)
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorSmoothingEffect.py]: Why this is needed?

File: Modules/Loadable/Segmentations/Resources/UI/qSlicerSegmentationsModule.ui [hunk 4]
```diff
@@ -7,7 +7,7 @@
     <x>0</x>
     <y>0</y>
     <width>348</width>
-    <height>719</height>
+    <height>844</height>
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/Resources/UI/qSlicerSegmentationsModule.ui]: this is a bit big... maybe collapse sections in the .ui file would reduce this?

Comment (Sunderlandkyl) [reply Modules/Loadable/Segmentations/Resources/UI/qSlicerSegmentationsModule.ui]: I've reduced the size as much as possible in the designer. If it is still too large, I can collapse some sections.

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorSmoothingEffect.py (hunk 3 above)
Comment (Sunderlandkyl) [reply Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorSmoothingEffect.py]: I investigated more, and the other overwrite mode changes in the AutoComplete effect were not required (after some bug fixing).

I've changed the overwrite mode to OverwriteVisibleSegments. I'm not sure if I understand how the layer mode would differ from the overwrite mode.

**Comment by lassoan** [comment]
Thank you, merged!
